File: 32-longest-valid-parentheses/32-longest-valid-parentheses.js
/**
 * @param {string} s
 * @return {number}
 */
var longestValidParentheses = function(s) {
    // trim leading ')'
    let l = 0;
    while (s[l] === ')') {
        l++;
    }
    
    // trim trailing '('
    let r = s.length;
    while (s[r - 1] === '(') {
        r--;
    }
    s = s.slice(l, r);
    
    let mirrored = s.split('').map(paren => paren === ')' ? '(' : ')').reverse().join('');
    
    const parse = s => {
        let [ans, stackHeight, validLen] = [0, 0, 0];

        for (let i = 0; i < s.length; i++) {
            if (s[i] === '(') {
                stackHeight++;
                validLen++;
            } else {
                stackHeight--;
                if (stackHeight < 0) {
                    validLen = 0;
                    stackHeight = 0;
                } else {
                    validLen++;
                }
                if (stackHeight === 0) ans = Math.max(ans, validLen);
            }
        }
        
        return ans;
    };
    
    return Math.max(parse(s), parse(mirrored));
};

File: 32-longest-valid-parentheses/32-longest-valid-parentheses.py
class Solution:
    def longestValidParentheses(self, s: str) -> int:
        if len(s) < 2:
            return 0
        dp = [0] * len(s)
        for i in range(len(s)):
            if s[i] == ')' and i-1 >= 0 and s[i-1] == '(':
                dp[i] = dp[i-2] + 2
            elif s[i] == ')' and i-1 - dp[i-1] >= 0 and s[i-1] == ')' and s[i-1 - dp[i-1]] == '(':
                dp[i] = dp[i-1] + dp[i-2 - dp[i-1]] + 2
        return max(dp)